src/qutip_qip/decompositions/general_decompositions.py

```diff
@@ -0,0 +1,134 @@
+#
+#    3. Neither the name of the QuTiP: Quantum Toolbox in Python nor the names
+#       of its contributors may be used to endorse or promote products derived
+#       from this software without specific prior written permission.
+#
+#    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+#    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###############################################################################
+
+import numpy as np
+
+def normalize_matrix(input_array)-> np.array:
+    """ Checks if the input gate's array is normalized or not. If not, makes
+    sure the input has been normalized.
+
+    This function will also check if the input is a valid array and a valid
+    quantum gate.
+
+     Args:
+        input_array : Matrix of a gate in Numpy array form.
+    """
+    if not isinstance(input_array, np.ndarray):
+        raise TypeError("Not a valid input : A Numpy input array must be provided.")
+
+
+    if isinstance(input_array, np.ndarray):
+        if input_array.size==0:
+            raise ValueError("An empty array was provided as input.")
+
+        if input_array.shape == (1,1):
+            raise ValueError("Provide a larger array as input.")
+
+        input_matrix_rows = input_array.shape[0]
+        input_matrix_columns = input_array.shape[1]
+        if input_matrix_rows !=input_matrix_columns:
+            raise ValueError("Input must be a square matrix to be a valid gate.")
+
+        if np.linalg.det(input_array) != 1:
+            if np.linalg.det(input_array) ==0:
+                raise ZeroDivisionError("Determinant of matrix =0.")
+            norm_factor = float(1/np.abs(np.linalg.det(input_array)))**0.5
+            input_array = np.around(norm_factor*input_array,5)
+        else:
+            input_array = input_array
+
+        return(input_array)
+
+# Note this function is defined for qobj, re-defined here for a numpy array.
+# Accessing individual elements of Qobj array could be problematic.
+def check_unitary(input_array)-> bool:
+    """Checks if the input matrix is unitary or not.
+
+    Args:
+       input_array : Matrix of a gate in Numpy array form.
+    """
+    try:
+        input_array = normalize_matrix(input_array)
+    except ZeroDivisionError:
+        input_array = input_array
```

Comment:
Don't think this exception has to be added. If this function's goal is to check if the matrix is unitary or not then a matrix whose determinant is zero is not a unitary matrix.

Reply:
Yeah you can return `False` already.

However, I'm not sure if normalizing is needed. It seems very unlikely to me that a not normalized gate will be unitary after the normalization. If someone provides something with `|det| != 1`, he/she already missed the point of providing a unitary.

Or is there a special reason?